packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp

```diff
@@ -0,0 +1,511 @@
+         class DVector,
+         class BVector,
+         class AMatrix,
+         class XVector,
+         class Scalar>
+static void
+chebyshev_kernel_vector
+(const Scalar& alpha,
+ const WVector& w,
+ const DVector& d,
+ const BVector& b,
+ const AMatrix& A,
+ const XVector& x,
+ const Scalar& beta)
+{
+  using execution_space = typename AMatrix::execution_space;
+
+  if (A.numRows () == 0) {
+    return;
+  }
+
+  int team_size = -1;
+  int vector_length = -1;
+  int64_t rows_per_thread = -1;
+
+  const int64_t rows_per_team =
+    chebyshev_kernel_vector_launch_parameters<execution_space>
+      (A.numRows (), A.nnz (), rows_per_thread, team_size, vector_length);
+  int64_t worksets = (b.extent (0) + rows_per_team - 1) / rows_per_team;
+
+  using Kokkos::Dynamic;
+  using Kokkos::Schedule;
+  using Kokkos::TeamPolicy;
+  using policy_type = TeamPolicy<execution_space, Schedule<Dynamic>>;
+  const char kernel_label[] = "chebyshev_kernel_vector";
+  policy_type policy (1, 1);
+  if (team_size < 0) {
+    policy = policy_type (worksets, Kokkos::AUTO, vector_length);
+  }
+  else {
+    policy = policy_type (worksets, team_size, vector_length);
+  }
+
+  // Canonicalize template arguments to avoid redundant instantiations.
+  using w_vec_type = typename WVector::non_const_type;
+  using d_vec_type = typename DVector::const_type;
+  using b_vec_type = typename BVector::const_type;
+  using matrix_type = AMatrix;
+  using x_vec_type = typename XVector::non_const_type;
+  using scalar_type = typename Kokkos::ArithTraits<Scalar>::val_type;
+
+  if (beta == Kokkos::ArithTraits<Scalar>::zero ()) {
+    constexpr bool use_beta = false;
+    using functor_type =
+      ChebyshevKernelVectorFunctor<w_vec_type, d_vec_type,
+                                        b_vec_type, matrix_type,
+                                        x_vec_type, scalar_type,
+                                        use_beta>;
+    functor_type func (alpha, w, d, b, A, x, beta, rows_per_team);
+    Kokkos::parallel_for (kernel_label, policy, func);
+  }
+  else {
+    constexpr bool use_beta = true;
+    using functor_type =
+      ChebyshevKernelVectorFunctor<w_vec_type, d_vec_type,
+                                        b_vec_type, matrix_type,
+                                        x_vec_type, scalar_type,
+                                        use_beta>;
+    functor_type func (alpha, w, d, b, A, x, beta, rows_per_team);
+    Kokkos::parallel_for (kernel_label, policy, func);
+  }
+}
+
+} // namespace Impl
+
+template<class TpetraOperatorType>
+ChebyshevKernel<TpetraOperatorType>::
+ChebyshevKernel (const Teuchos::RCP<const operator_type>& A)
+{
+  setMatrix (A);
+}
+
+template<class TpetraOperatorType>
+void
+ChebyshevKernel<TpetraOperatorType>::
+setMatrix (const Teuchos::RCP<const operator_type>& A)
+{
+  if (A_op_.get () != A.get ()) {
+    A_op_ = A;
+
+    // We'll (re)allocate these on demand.
+    X_colMap_ = std::unique_ptr<vector_type> (nullptr);
```

Comment:
I wish there were a way to get at the column-Map MultiVector that already lives in CrsMatrix.